ProgressTracker should have a WeakPtr instead of a RefPtr<LocalFrame>

achristensen07 opened the pull request:
#### 3cc50bc287a862a00f885e5699d6377d9a1c236c
<pre>
ProgressTracker should have a WeakPtr instead of a RefPtr&lt;LocalFrame&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=285649">https://bugs.webkit.org/show_bug.cgi?id=285649</a>
<a href="https://rdar.apple.com/142492439">rdar://142492439</a>

Reviewed by Charlie Wolfe.

There are common crashes resulting from the ProgressTracker destructor
destroying a Frame.  The ProgressTracker does not need to keep the Frame
alive, so use a WeakPtr instead.

Since it can now become null at any point, I added null checks before
using it everywhere.  Since it no longer keeps the LocalFrame alive,
I create a RefPtr on the stack before using it.

* Source/WebCore/loader/ProgressTracker.cpp:
(WebCore::ProgressTracker::progressStarted):
(WebCore::ProgressTracker::finalProgressComplete):
(WebCore::ProgressTracker::incrementProgress):
(WebCore::ProgressTracker::progressHeartbeatTimerFired):
* Source/WebCore/loader/ProgressTracker.h:

Canonical link: <a href="https://commits.webkit.org/288634@main">https://commits.webkit.org/288634@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dac53c458ee8895bdc26418f51d1c529e03d6298

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83997 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3615 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38298 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89072 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35005 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86082 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3706 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11583 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65329 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23170 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87043 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/2761 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76321 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45622 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2693 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30544 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34054 "Built successfully") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73656 "Build is in progress. Recent messages:") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31302 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90450 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11262 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8150 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73780 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11486 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72150 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72997 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17287 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2601 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12992 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11214 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16686 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11062 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14538 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12834 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebCore/loader/ProgressTracker.cpp:
```diff
@@ -121,7 +121,7 @@ void ProgressTracker::progressStarted(LocalFrame& frame)
         m_progressHeartbeatTimer.startRepeating(progressHeartbeatInterval);
-        RefPtr originatingProgressFrame = m_originatingProgressFrame;
+        RefPtr originatingProgressFrame = m_originatingProgressFrame.get();
         originatingProgressFrame->protectedLoader()->loadProgressingStatusChanged();
```
Should probably null check this now?

achristensen07 replied:
And other places!

charliewolfe commented on Source/WebCore/loader/ProgressTracker.cpp:
```diff
@@ -121,21 +121,23 @@ void ProgressTracker::progressStarted(LocalFrame& frame)
         m_progressHeartbeatTimer.startRepeating(progressHeartbeatInterval);
-        RefPtr originatingProgressFrame = m_originatingProgressFrame;
-        originatingProgressFrame->protectedLoader()->loadProgressingStatusChanged();
+        if (RefPtr originatingProgressFrame = m_originatingProgressFrame.get()) {
```
Oh, null check isn’t _actually_ necessary here, since we have `m_originatingProgressFrame = &frame;` above.

charliewolfe commented on Source/WebCore/loader/ProgressTracker.cpp:
```diff
@@ -149,27 +151,27 @@ void ProgressTracker::progressEstimateChanged(LocalFrame& frame)
-    PROGRESS_TRACKER_RELEASE_LOG(PROGRESSTRACKER_PROGRESSCOMPLETED, frame.frameID().object().toUInt64(), m_progressValue, m_numProgressTrackedFrames, m_originatingProgressFrame ? m_originatingProgressFrame->frameID().object().toUInt64() : 0, m_isMainLoad);
+    RefPtr originatingProgressFrame = m_originatingProgressFrame.get();
+    LOG(Progress, "Progress completed (%p) - frame %p(frameID %" PRIu64 "), value %f, tracked frames %d, originating frame %p", this, &frame, frame.frameID().object().toUInt64(), m_progressValue, m_numProgressTrackedFrames, originatingProgressFrame.get());
+    PROGRESS_TRACKER_RELEASE_LOG(PROGRESSTRACKER_PROGRESSCOMPLETED, frame.frameID().object().toUInt64(), m_progressValue, m_numProgressTrackedFrames, originatingProgressFrame ? originatingProgressFrame->frameID().object().toUInt64() : 0, m_isMainLoad);
```
Doesn’t matter either way, but smart pointer guidelines say we don’t need smart pointers on the stack when objects only call trivial and inlined functions, which is the case here.

charliewolfe commented on Source/WebCore/loader/ProgressTracker.h:
```diff
@@ -34,6 +34,7 @@
 #include <wtf/Noncopyable.h>
 #include <wtf/RefPtr.h>
```
Could remove this include now

achristensen07 committed:
ProgressTracker should have a WeakPtr instead of a RefPtr<LocalFrame>
https://bugs.webkit.org/show_bug.cgi?id=285649
rdar://142492439

Reviewed by Charlie Wolfe.

There are common crashes resulting from the ProgressTracker destructor
destroying a Frame.  The ProgressTracker does not need to keep the Frame
alive, so use a WeakPtr instead.

Since it can now become null at any point, I added null checks before
using it everywhere.  Since it no longer keeps the LocalFrame alive,
I create a RefPtr on the stack before using it.

* Source/WebCore/loader/ProgressTracker.cpp:
(WebCore::ProgressTracker::progressStarted):
(WebCore::ProgressTracker::finalProgressComplete):
(WebCore::ProgressTracker::incrementProgress):
(WebCore::ProgressTracker::progressHeartbeatTimerFired):
* Source/WebCore/loader/ProgressTracker.h:

Canonical link: https://commits.webkit.org/288634@main
